pynetgear/__init__.py

```diff
@@ -64,32 +64,40 @@ def get_attached_devices(self):
         if not success:
             return None
 
-        data = re.search(r"<NewAttachDevice>(.*)</NewAttachDevice>",
-                         response).group(1).split(";")
+        raw = re.search(REGEX_ATTACHED_DEVICES, response).group(1)
 
+        # Netgear inserts a double-encoded value for "unknown" devices
+        decoded = raw.replace(UNKNOWN_DEVICE_ENCODED, UNKNOWN_DEVICE_DECODED)
+
+        entries = data.split("@")
         devices = []
 
-        device_start = [index for index, value in enumerate(data)
-                        if '@' in value]
+        # First element is the total device count
+        if entries:
+            entry_count = int(entries.pop(0))
+        else:
+            _LOGGER.error("Error parsing device-list")
+            return devices
+
+        if(entry_count != len(entries)):
+            _LOGGER.warning(
+                """Number of devices should \
+                 be: %d but is: %d""", entry_count, len(entries))
 
-        for index, start in enumerate(device_start):
-            try:
-                info = data[start:device_start[index+1]]
-            except IndexError:
-                # The last device, ignore the last element
-                info = data[start:-1]
+        for entry in entries:
+            info = entry.split(";")
 
             if len(info) == 0:
                 continue
-            elif len(info) < 4:
-                _LOGGER.warning('Unexpected entry: %s', info)
+            elif len(info) != (4 or 6):
```

Comment:
Resolves to `true` for routers with link-type & rate. 
Should be `elif len(info) != (4 or 7):`